### PR TITLE
[FC-91] Updated Cheddar to support Agents

### PR DESCRIPTION
**Pull request by clicktravel-basharat**
@ClickTravel/cheddarchapter

This is a *non-breaking* change as I have added deprecated methods to allow the previous ```getPrincipal()```, ```setPrincipal()``` and ```clearPrincipal()``` methods to work.

There is still some uncertainly (need discussion from API Chapter) over the choice of HTTP Headers to contain the principal/agent ids in the ```ContainerSecurityRequestFilter``` class, but the rest can still be reviewed.

### CODE REVIEW

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/AgentSecurityContext.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.cheddar.request.context;
+
+public class AgentSecurityContext implements SecurityContext {
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/AgentSecurityContext.java]: This class should have Javadoc explaining the concepts of impersonation and define the term 'agent'.

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/BasicSecurityContext.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.cheddar.request.context;
+
+public class BasicSecurityContext implements SecurityContext {
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/BasicSecurityContext.java]: Maybe helpful to Javadoc this class and say that this class is used when no impersonation is occurring. Perhaps add a ```@see AgentSecurityContext```

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/SecurityContext.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.cheddar.request.context;
+
+public interface SecurityContext {
+
+    String principal();
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/SecurityContext.java]: This needs some Javadoc to define what ```principal()``` returns, and to make it clear how impersonation affects this.

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat First review complete on work so far. Seems OK but needs some Javadoc in the places I have commented, see above.

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/SecurityContext.java (hunk 3 above)
Comment (clicktravel-basharat) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/SecurityContext.java]: I don't think that the interface should define *how* the implementing classes determine principal. Would it be sufficient to say what the principal is (as taken from online defintion), then list the known implementations?

**Comment by clicktravel-basharat** [comment]
@clicktravel-steffan Javadoc added

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/BasicSecurityContext.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.cheddar.request.context;
+
+/**
+ * Obtains the currently authenticated principal's user id when the principal acts on their own behalf without an agent.
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/BasicSecurityContext.java]: Suggested clarified Javadoc text, aligned with comment above:
```Identifies the currently authenticated user as the principal in the security context. This is used when the user acts on their own behalf without an agent.```

**Comment by clicktravel-basharat** [comment]
@clicktravel-steffan Javadoc update

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/AgentSecurityContext.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package com.clicktravel.cheddar.request.context;
+
+/**
+ * Obtains the currently authenticated principal's user id when an agent acts on behalf of the authenticated user.
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/AgentSecurityContext.java]: Suggested clarified Javadoc text, see also comment below:
```Identifies an effective principal as the principal in the security context, and a supporting agent principal. This is used when an agent acts on behalf of (i.e. impersonates) a supported user.```

**Comment by clicktravel-basharat** [comment]
@clicktravel-steffan Javadoc update

File: cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/SecurityContext.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+package com.clicktravel.cheddar.request.context;
+
+/**
+ * Interface defining the minimum security information associated with the current thread of execution.
+ *
+ * @see {@link BasicSecurityContext} and {@link AgentSecurityContext}
+ */
+public interface SecurityContext {
+
+    String principal();
```
Comment (clicktravel-steffan) [inline cheddar/cheddar-context/src/main/java/com/clicktravel/cheddar/request/context/SecurityContext.java]: Add Javadoc here:
```@return The user ID of the principal in the security context. Domain actions (commands or queries) are performed 'as' this user. This is the actor which any authorisation checks should be performed against.```

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat I've added my suggestions for Javadoc, please see comments above.

**Comment by clicktravel-steffan** [comment]
@clicktravel-basharat OK, all updates are now fine. Good to merge.

**Pull request merged**